pytest_inmanta/plugin.py

```diff
@@ -951,6 +951,39 @@ def dryrun_resource(
         assert ctx.status == status
         return ctx.changes
 
+    def dryrun_all(
+        self, run_as_root: bool = False, assert_create_or_delete: bool = False
+    ) -> Dict[Resource, HandlerContext]:
+        """
+        Runs a dryrun for every resource.
+        :param run_as_root: run the mock agent as root
+        :param assert_create_or_delete: assert that every resource will either be created or deleted.
+        """
+        results = {}
+        for resource in self.resources.values():
+            ctx = self.dryrun(resource, run_as_root=run_as_root)
+            if ctx.changes:
+                results[resource] = ctx
+            if assert_create_or_delete:
+                assert "purged" in ctx.changes
+        return results
+
+    def dryrun_and_deploy_all(
+        self, run_as_root: bool = False, assert_create_or_delete: bool = False
+    ) -> Dict[str, Dict[Resource, HandlerContext]]:
+        """
+        Runs a dryrun, followed by a deploy and a final dryrun for every resource and asserts the expected behaviour.
+        :param run_as_root: run the mock agent as root
+        :param assert_create_or_delete: assert that every resource will either be created or deleted.
+        """
+        dryrun_result = self.dryrun_all(
+            run_as_root=run_as_root, assert_create_or_delete=assert_create_or_delete
+        )
+        deploy_result = self.deploy_all(run_as_root=run_as_root)
+
+        assert not self.dryrun_all(run_as_root=run_as_root)
+        return {"dryrun": dryrun_result, "deploy": deploy_result.results}
```

Comment:
We generally try to avoid untyped dicts for structured data. You could use a `TypedDict` but I would suggest introducing a dataclass for this result. It could also help to add some helper methods, so the end user can inspect state like e.g. `assert dry_run_and_deploy_all().all_ok()` (naming might not be good but I think it gets the point accross).

Reply:
For new things it should be dataclass for better typing of existing code use TypedDict